### ISSUE TITLE
[Bug] AdamW weight decay 在 Adam update 之後才套用，順序錯誤導致與 PyTorch 行為不同

### ISSUE THREAD

**Issue by s9213712** [s9213712/minicnn#39]
## 問題描述

`src/minicnn/optim/adamw.py` 中 weight decay 套用在 Adam update 之後：

```python
def step(self):
    for i, p in enumerate(self.params):
        ...
        # 步驟 1：Adam 梯度更新
        p.data = p.data - self.lr * m_hat / (np.sqrt(v_hat) + self.eps)
        # 步驟 2：Weight decay 在已更新的 p.data 上套用！
        if self.weight_decay:
            p.data = p.data - self.lr * self.weight_decay * p.data  # ← 使用已更新的 p.data
```

## 正確的 AdamW 公式

根據 Loshchilov & Hutter (2019) 原論文，AdamW 的正確更新順序應為：
```
w_{t+1} = (1 - η*λ) * w_t - η * m̂_t / (√v̂_t + ε)
```

即 weight decay 作用在**更新前**的參數 `w_t`，而非更新後的 `w_{t+1}`。

等效的正確實作：
```python
# 先計算 weight decay 貢獻（作用在更新前的 w_t）
wd_update = self.lr * self.weight_decay * p.data  # 使用原始 p.data
# 再做 Adam 更新
p.data = p.data - self.lr * m_hat / (np.sqrt(v_hat) + self.eps) - wd_update
```

## 影響

- **嚴重性**: Major
- 使用 `weight_decay > 0` 的 AdamW 訓練結果與 PyTorch 基線不可比較
- 在大 `lr * weight_decay` 時差異明顯（例如 lr=1e-3, wd=0.01 時每步額外損失 ~0.001% 準確度）

## 相關位置

- `src/minicnn/optim/adamw.py:42-44`
